src/pkg/extensions/extensions.go

```diff
@@ -0,0 +1,126 @@
+package extensions
+
+import (
+	"context"
+	"io"
+
+	"github.com/alcionai/clues"
+
+	"github.com/alcionai/corso/src/pkg/backup/details"
+	"github.com/alcionai/corso/src/pkg/logger"
+)
+
+// Extension client interface
+type CorsoItemExtension interface {
+	io.ReadCloser
+}
+
+type CorsoItemExtensionFactory func(
+	context.Context,
+	io.ReadCloser,
+	details.ItemInfo,
+	*details.ExtensionInfo,
+) (CorsoItemExtension, error)
+
+// Thin wrapper for runtime logging & metrics around extensions
+type loggerExtension struct {
+	info    details.ItemInfo
+	innerRc io.ReadCloser
+	ctx     context.Context
+	extInfo *details.ExtensionInfo
+}
+
+func NewLoggerExtension(
+	ctx context.Context,
+	rc io.ReadCloser,
+	info details.ItemInfo,
+	extInfo *details.ExtensionInfo,
+) (CorsoItemExtension, error) {
+	return &loggerExtension{
+		ctx:     ctx,
+		innerRc: rc,
+		info:    info,
+		extInfo: extInfo,
+	}, nil
+}
+
+func (l *loggerExtension) Read(p []byte) (int, error) {
+	n, err := l.innerRc.Read(p)
+	if err != nil && err != io.EOF {
+		logger.CtxErr(l.ctx, err).Error("inner read")
+		return n, err
+	}
+
+	if err == io.EOF {
+		logger.Ctx(l.ctx).Debug("corso extensions: EOF")
+	}
+
+	return n, err
+}
+
+func (l *loggerExtension) Close() error {
+	err := l.innerRc.Close()
+	if err != nil {
+		logger.CtxErr(l.ctx, err).Error("inner close")
+		return err
+	}
+
+	logger.Ctx(l.ctx).Info("corso extensions: closed")
+
+	return nil
+}
+
+type AddItemExtensioner interface {
+	AddItemExtensions(
+		context.Context,
+		io.ReadCloser,
+		details.ItemInfo,
+		[]CorsoItemExtensionFactory,
+	) (io.ReadCloser, *details.ExtensionInfo, error)
+}
+
+var _ AddItemExtensioner = &ItemExtensionHandler{}
+
+type ItemExtensionHandler struct{}
+
+// AddItemExtensions wraps provided readcloser with extensions
+// supplied via factory, with the first extension in slice being
+// the innermost one.
+func (eh *ItemExtensionHandler) AddItemExtensions(
+	ctx context.Context,
+	rc io.ReadCloser,
+	info details.ItemInfo,
+	factories []CorsoItemExtensionFactory,
+) (io.ReadCloser, *details.ExtensionInfo, error) {
+	if rc == nil {
+		return nil, nil, clues.New("nil inner readcloser")
+	}
+
+	if len(factories) == 0 {
+		return nil, nil, clues.New("no extensions supplied")
```

Comment:
My thought process was that if the caller called this function, they have 100% intent to add extensions. So we should fail in caller rather than silently skipping extensions.

What do you think?

Reply:
The way I read the code, it seems that we'd be the ones instantiating the call, not the extension provider.  In that case the ball is in our court, and a no-op here saves us a len conditional.